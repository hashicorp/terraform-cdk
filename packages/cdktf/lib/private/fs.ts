// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { execSync } from "child_process";
import * as fs from "fs";
import * as path from "path";
import * as crypto from "crypto";

const HASH_LEN = 32;

// Full implementation at https://github.com/jprichardson/node-fs-extra/blob/master/lib/copy/copy-sync.js
/**
 * Copy a file or directory. The directory can have contents and subfolders.
 * @param {string} src
 * @param {string} dest
 */
export function copySync(src: string, dest: string) {
  /**
   * Copies file if present otherwise walks subfolder
   * @param {string} p
   */
  function copyItem(p: string) {
    const sourcePath = path.resolve(src, p);
    const stat = fs.statSync(sourcePath);
    if (stat.isFile()) {
      fs.copyFileSync(sourcePath, path.resolve(dest, p));
    }
    if (stat.isDirectory()) {
      walkSubfolder(p);
    }
  }
  /**
   * Copies contents of subfolder
   * @param {string} p
   */
  function walkSubfolder(p: string) {
    const sourceDir = path.resolve(src, p);
    fs.mkdirSync(path.resolve(dest, p), { recursive: true });
    fs.readdirSync(sourceDir).forEach((item) => copyItem(path.join(p, item)));
  }

  walkSubfolder(".");
}

/**
 * Zips contents at src and places zip archive at dest
 * @param {string} src
 * @param {string} dest
 */
export function archiveSync(src: string, dest: string) {
  const projectRoot = path.resolve(__dirname, "..", "..");
  const zipSyncPath = path.resolve(projectRoot, "bin", "zipSync.js");
  execSync(`node ${zipSyncPath} ${src} ${dest}`);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function hashPath(src: string) {
  const hash = crypto.createHash("md5");

  // eslint-disable-next-line jsdoc/require-jsdoc
  function hashRecursion(p: string) {
    const stat = fs.statSync(p);
    if (stat.isFile()) {
      hash.update(fs.readFileSync(p));
    } else if (stat.isDirectory()) {
      fs.readdirSync(p).forEach((filename) =>
        hashRecursion(path.resolve(p, filename))
      );
    }
  }

  hashRecursion(src);
  return hash.digest("hex").slice(0, HASH_LEN).toUpperCase();
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function findFileAboveCwd(
  file: string,
  rootPath = process.cwd()
): string | null {
  const fullPath = path.resolve(rootPath, file);
  if (fs.existsSync(fullPath)) {
    return fullPath;
  }

  const parentDir = path.resolve(rootPath, "..");
  if (fs.existsSync(parentDir) && parentDir !== rootPath) {
    return findFileAboveCwd(file, parentDir);
  }

  return null;
}
