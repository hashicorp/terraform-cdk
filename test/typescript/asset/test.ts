// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { TestDriver } from "../../test-helper";
import * as path from "path";
import * as fs from "fs";

describe("full integration test", () => {
  let driver: TestDriver;

  beforeAll(async () => {
    driver = new TestDriver(__dirname);
    await driver.setupTypescriptProject();
    driver.copyFiles("local-asset.txt", "relative-asset.txt");
    driver.copyFolders("fixtures", "relative");
  });

  test("synth", async () => {
    await driver.synth("fixed");
    expect(driver.synthesizedStack("fixed").toString()).toMatchSnapshot();
  });

  test("file asset copied", async () => {
    await driver.synth("fixed");
    expect(
      fs.readFileSync(
        path.resolve(
          driver.stackDirectory("fixed"),
          "assets/local-asset/hash/local-asset.txt"
        ),
        "utf-8"
      )
    ).toMatchSnapshot();
  });

  test("folder asset copied", async () => {
    await driver.synth("fixed");
    expect(
      fs.readFileSync(
        path.resolve(
          driver.stackDirectory("fixed"),
          "assets/fixtures/hash/a.txt"
        ),
        "utf-8"
      )
    ).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.resolve(
          driver.stackDirectory("fixed"),
          "assets/fixtures/hash/b.txt"
        ),
        "utf-8"
      )
    ).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.resolve(
          driver.stackDirectory("fixed"),
          "assets/fixtures/hash/foo/bar/c.txt"
        ),
        "utf-8"
      )
    ).toMatchSnapshot();
  });

  test("zip file created", async () => {
    await driver.synth("fixed");

    const stat = fs.statSync(
      path.resolve(
        driver.stackDirectory("fixed"),
        "assets/zipped-fixtures/hash/archive.zip"
      )
    );
    expect(stat.isFile()).toBe(true);
  });

  test("relative file asset copied", async () => {
    await driver.synth("fixed");
    expect(
      fs.readFileSync(
        path.resolve(
          driver.stackDirectory("fixed"),
          "assets/relative-asset/hash/relative-asset.txt"
        ),
        "utf-8"
      )
    ).toMatchSnapshot();
  });

  test("relative folder asset copied", async () => {
    await driver.synth("fixed");
    expect(
      fs.readFileSync(
        path.resolve(
          driver.stackDirectory("fixed"),
          "assets/relative/hash/a.txt"
        ),
        "utf-8"
      )
    ).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.resolve(
          driver.stackDirectory("fixed"),
          "assets/relative/hash/b.txt"
        ),
        "utf-8"
      )
    ).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.resolve(
          driver.stackDirectory("fixed"),
          "assets/relative/hash/bar/c.txt"
        ),
        "utf-8"
      )
    ).toMatchSnapshot();
  });

  test("without asset changes there should be no redeployment", async () => {
    expect(await driver.diff("normal")).toContain(
      "1 to add, 0 to change, 0 to destroy"
    );
    await driver.deploy(["normal"]);
    expect(await driver.diff("normal")).toContain("No changes");
  });
});
