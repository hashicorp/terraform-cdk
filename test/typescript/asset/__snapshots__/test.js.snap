// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`full integration test file asset copied 1`] = `"Hello Asset World"`;

exports[`full integration test folder asset copied 1`] = `"This is a"`;

exports[`full integration test folder asset copied 2`] = `"This is b"`;

exports[`full integration test folder asset copied 3`] = `"this is c"`;

exports[`full integration test synth 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"fixed\\",
      \\"backend\\": \\"local\\"
    }
  },
  \\"output\\": {
    \\"localassetname\\": {
      \\"value\\": \\"local-asset.txt\\"
    },
    \\"fixturesname\\": {
      \\"value\\": \\"fixtures\\"
    },
    \\"zippedfixturesname\\": {
      \\"value\\": \\"archive.zip\\"
    }
  }
}"
`;
